### DISCUSSION TITLE
Linux Build

### DISCUSSION THREAD

**Discussion by VinnyVicious** [o3de/o3de#1884] [Q&A]
Are there instructions on how to build the engine for Linux? Not asking about the editor, just the engine runtime.

**Comment by daniel-nth** [comment]
There are open issues for this. 
Editor #746
Client #745

**Comment by amznestebanpapp** [comment]
Its possible to build the engine for Linux, however, the "client" (engine runtime) and the Editor needs some work (as those issues point out).
The main part that should be running is the server side. Asset processing is running (we validate that in Jenkins). 

As you can tell, we don't want to say that is "working" without validating, filling the gaps and running some QA on it. Since this is open source, you can give it a try and see how far you can get.

If you are interested, I can give you the build instructions. Join the discord sig-build channel, there are partners building it in Linux (even in ARM)

**Reply by OBWANDO** [reply]
Can you also post the build steps in here as well. I'd like to use it as an artifact for an RFC.

**Reply by amznestebanpapp** [reply]
For Ubuntu 18.04 we had:
Setup
```
$ sudo apt-get update
$ sudo apt-get upgrade
$ sudo apt-get autoremove
$ sudo apt-get install build-essential
$ sudo apt-get install clang-6.0 uuid-dev libz-dev libncurses5-dev libcurl4-openssl-dev
$ sudo apt-get install mesa-common-dev libglu1-mesa-dev
$ sudo apt-get install libjpeg-dev libjbig-dev uuid-dev libsdl2-dev
$ sudo apt-get install libxcb-xinerama0 libxcb-xinput0
```
Generation
```
cmake -B build/linux -S <source-dir> -G "Ninja Multi-Config" -DCMAKE_C_COMPILER=clang-6.0 -DCMAKE_CXX_COMPILER=clang++-6.0 -DLY_3RDPARTY_PATH=<3rdPartyPath> -DLY_PROJECTS=AutomatedTesting
```
Build
```
cmake --build build/linux --config profile --target all
```

**Reply by jspanchu** [reply]
I got it to configure, generate and build on latest arch linux 5.12.15 with clang-12.0.1.

Had to install libffi6 from aur to get the python runtime built. 

Running the editor shows a  failed symbol lookup in libssh2

```
[jaswantp@nixboxjp o3de-clang-12.0.0-x86_64-pc-linux-gnu-Debug]$ lldb ./bin/debug/Editor
(lldb) target create "./bin/debug/Editor"
Current executable set to '/home/jaswantp/builds/o3de-clang-12.0.0-x86_64-pc-linux-gnu-Debug/bin/debug/Editor' (x86_64).
(lldb) r
Process 67865 launched: '/home/jaswantp/builds/o3de-clang-12.0.0-x86_64-pc-linux-gnu-Debug/bin/debug/Editor' (x86_64)
ProjectManager: 
==================================================================
ProjectManager: Trace::Error
 /home/jaswantp/repos/o3de/Code/Framework/AzFramework/AzFramework/ProjectManager/ProjectManager.cpp(81): 'AzFramework::ProjectManager::ProjectPathCheckResult AzFramework::ProjectManager::CheckProjectPathProvided(const int, char **)'
ProjectManager: Project Manager failed to launch and no project selected!
ProjectManager: ==================================================================
Module: Attempting to load module:/home/jaswantp/builds/o3de-clang-12.0.0-x86_64-pc-linux-gnu-Debug/bin/debug/libEditorLib.so
Module: Failed with error:
/usr/lib/libssh2.so.1: undefined symbol: EVP_cast5_cbc, version OPENSSL_1_1_0
System: 
==================================================================
System: Trace::Assert
 /home/jaswantp/repos/o3de/Code/Editor/main.cpp(26): (140737331076160) 'int main(int, char **)'
System: EditorLib could not be loaded
System: ------------------------------------------------
System: main (+0xe3) [0x5555555f6313]
System: __libc_start_main (+0xd5) [0x7ffff6a2db25]
System: _start (+0x2e) [0x5555555f615e]
System: ==================================================================
We should not delete an environment while there are 1 modules attached! Unload all DLLs first!Process 67865 exited with status = 1 (0x00000001) 
(lldb) 
```

**Reply by Goddard** [reply]
Is that the correct path for your libssh?

I'm on Ubuntu so can't compare, but mine is at /lib/x86_64-linux-gnu/libssh2.so.1

**Reply by jspanchu** [reply]
> Is that the correct path for your libssh?
> 
> I'm on Ubuntu so can't compare, but mine is at /lib/x86_64-linux-gnu/libssh2.so.1

```bash
$ ls /usr/lib | grep libssh
libssh2.so
libssh2.so.1
libssh2.so.1.0.1
libssh.so
libssh.so.4
libssh.so.4.8.6
```

**Reply by Goddard** [reply]
I assume you ran sudo ldconfig?

**Reply by jspanchu** [reply]
> I assume you ran sudo ldconfig?

No. Should I? Does the build process alter /etc/ld.so.conf? I don't think so.

It's the same even after sudo ldconfig.

**Reply by jspanchu** [reply]
After some digging around on a Sunday, there appears to be a conflict here. Pretty common when libraries try to embed 3rd party binaries in their own build tree.

```
libEditorLib.so depends on system's libssh2.so # I assume this comes from AzNetworking.
libEditorLib.so depends on libopensl.so.
```
Conflict:
On an arch linux system, the system's `libssh2.so` depends on system's `libopenssl.so` (v1.1.1k) whereas `libEditorLib.so` links to 3rd party openssl-`v1.1.1b` (downloaded by ly_associate_package).

I'll try to let CMake look up the system's OpenSSL instead. We must teach ly/o3de's CMake to use system/external packages if the editor is ever gonna be packaged on Linux distros (unless the plan is containerized app images! I think unity does this)

**Comment by pyramid3d** [comment]
I was able to build and run o3de on Linux by following the excellent 
[Building for Linux](https://o3de.org/docs/welcome-guide/setup/setup-from-github/building-linux/) docu.

What I missed though was a command line to install the build. Tried a couple of iterations of:
```bash
> cmake --install build/linux/bin/ --prefix /opt/o3de/
> cmake --install build/linux/ --prefix /opt/o3de
? ...
```
... to no avail.

What would be the correct command line to install the build?

**Reply by Zaya6** [reply]
How'd you get passed :
`cmake --build build/linux --target Editor --config profile -j 8` on that guide?

All I get is unknown target. Everything works on every other step until then

Though there are alternate build instructions I just found here https://o3de.org/docs/user-guide/build/

I'll try that today

**Reply by pyramid3d** [reply]
I haven't tried the ninja build. Was using Unix only:
```bash
> cmake -B build/linux -S . -G "Unix Makefiles" -DLY_3RDPARTY_PATH=../o3de-packages
> cmake --build build/linux --target Editor --config profile -j 8
```

**Reply by matteogrs** [reply]
To install the engine as a pre-built SDK, you have to pass the `install` target in your second command. No need for `--config` option with makefile, since it is a single-config build:
```
cmake --build build/linux --target install
```

An alternative way to install the engine is to use the following unofficial tool, which wraps all the required checks, commands and options into single mnemonic commands (`install engine`, `init project`, `open`, ...). Choose `native mode` if you don't need an isolated environment based on containers: https://github.com/loherangrin/o3tanks

**Reply by amznestebanpapp** [reply]
The error you are likely getting is around permissions. When setting the prefix to `/opt` you need to be root to write there. You can pre-create the folder `/opt/o3de` and give rw access to your user/group. 
If errors are not related to permissions, can you post them?

**Reply by pyramid3d** [reply]
@loherangrin @amznestebanpapp ,
Thanks for your both replies. I tried a clean "Unix Makefiles" build with:
```
cmake --build build/linux --target install
```

It results in the following error (I have write permissions to /opt):
```bash
Install the project...
-- Install configuration: "profile"
CMake Error at Registry/cmake_install.cmake:53 (file):
  file INSTALL cannot find
  "/home/data/projects/pyramid/Android/o3de.git/build/linux/bin/debug/Registry":
  No such file or directory.
```

The strange thing is, the installation is trying to look in bin/debug directory which does not exist (hardcoded instead of profile?).

**Reply by amznestebanpapp** [reply]
That looks like you are generating in a non-multi config generator (Makefile/Ninja) without passing the configuration.
The "Ninja" generator is a single-configuration generator, "Ninja Multi-Config" is the multi-config one. Makefiles is a singe-configurator generator. 
When dealing with a single-configurator generator, you need to pass the configuration you are working with using [CMAKE_BUILD_TYPE](https://cmake.org/cmake/help/latest/variable/CMAKE_BUILD_TYPE.html#variable:CMAKE_BUILD_TYPE). To do, pass `-DCMAKE_BUILD_TYPE=profile`

**Reply by pyramid3d** [reply]
Since upgrade to clang-13 I am kinda stuck with this error on the latest pull:
```
In file included from /mnt/data/projects/pyramid/Android/o3de.git/build/linux/External/Atom-99ca8ee7/RPI/Code/CMakeFiles/Atom_RPI.Public.dir/Unity/unity_8_cxx.cxx:13:
/mnt/data/projects/pyramid/Android/o3de.git/Gems/Atom/RPI/Code/Source/RPI.Public/Culling.cpp:315:22: error: variable 'numDrawPackets' set but not used [-Werror,-Wunused-but-set-variable]
            uint32_t numDrawPackets = 0;
```

FYI, I have rebuilt an earlier commit successfully with clang-13:
```
git checkout 'development@{2021-12-04 23:30:00}'
```

**Reply by amznestebanpapp** [reply]
> Since upgrade to clang 13 I am kinda stuck with this error:
> 
> ```
> In file included from /mnt/data/projects/pyramid/Android/o3de.git/build/linux/External/Atom-99ca8ee7/RPI/Code/CMakeFiles/Atom_RPI.Public.dir/Unity/unity_8_cxx.cxx:13:
> /mnt/data/projects/pyramid/Android/o3de.git/Gems/Atom/RPI/Code/Source/RPI.Public/Culling.cpp:315:22: error: variable 'numDrawPackets' set but not used [-Werror,-Wunused-but-set-variable]
>             uint32_t numDrawPackets = 0;
> ```

seems to be an unused variable, just remove it
Compilers over time (newer versions) get better at detecting stuff, so it is expected that a newer compiler will find more things. Feel free to fix it and put a PR.
